crates/wasm-smith/src/lib.rs

```diff
@@ -550,25 +584,137 @@ where
         Ok(())
     }
 
-    fn arbitrary_types(&mut self, u: &mut Unstructured) -> Result<()> {
-        arbitrary_loop(u, self.config.min_types(), self.config.max_types(), |u| {
-            let mut params = vec![];
-            let mut results = vec![];
-            arbitrary_loop(u, 0, 20, |u| {
-                params.push(self.arbitrary_valtype(u)?);
-                Ok(true)
-            })?;
-            arbitrary_loop(u, 0, 20, |u| {
-                results.push(self.arbitrary_valtype(u)?);
-                Ok(true)
-            })?;
-            self.types.push(FuncType { params, results });
+    fn arbitrary_initializers(&mut self, u: &mut Unstructured) -> Result<()> {
+        if self.config.module_linking_enabled() {
+            loop {
+                match u.int_in_range(0..=2)? {
+                    0 => self.arbitrary_types(0, u)?,
+                    1 => self.arbitrary_imports(0, u)?,
+                    _ => break,
+                }
+            }
+            // If after generating a list of sections we haven't met our
+            // minimum quotas then meet them now.
+            if let Some(min) = self.config.min_types().checked_sub(self.types.len()) {
+                if min > 0 {
+                    self.arbitrary_types(min, u)?;
+                }
+            }
+            if let Some(min) = self.config.min_imports().checked_sub(self.imports.len()) {
+                if min > 0 {
+                    self.arbitrary_imports(min, u)?;
```

Comment:
We can simultaneously get rid of a couple lines and make it read clearer (IMO):

```suggestion
            if self.imports.len() < self.config.min_imports() {
                self.arbitrary_imports(self.config.min_imports() - self.imports.len(), u)?;
```